[3-4단계 - 방탈출 예약 대기] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/4374

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

안녕하세요 베루스~ 2단계 제출합니다!

#### 이전 단계에서 주신 피드백에 대한 답변
> 빈으로 등록된 객체들은 대부분 상태가 변경되지 않는 객체 아닌가요? 컨텍스트 격리가 필요하다면 @DirtiesContext를 사용하는게 맞지만, 지금 구현에서는 필요한지 의문입니다.

이번 단계를 진행하며 여러 방면으로 확인해보았고 현재는 베루스의 의견에 동의합니다!
그동안에는 `@DirtiesContext`를 사용하지 않으면 일부 테스트가 실패했어서 명시적으로 `@DirtiesContext`를 사용했었습니다.
확인 결과 일부 테스트에서 컨텍스트를 공유하고, 그 과정에서 스크립트로 초기화한 데이터들이 남아있어 테스트 간의 독립성이 보장되지 않았었던 것이었습니다.
말씀주신 것처럼 현재 구현에서는 스프링 빈들의 상태가 변경되지 않기 때문에, 컨텍스트를 새로 로드하기보다는 각 테스트 별로 명시적 DB 초기화를 해주어 독립성을 보장했습니다.

> 결국 서비스에서는 DB 접근과 관련된 기술을 MemberRepository 인터페이스 내부에 숨겨졌기 때문에 관심사가 분리되겠네요~

직접 구조를 그려 설명해주셔서 감사드립니다.👍 덕분에 어떤 방식인지 확실히 이해했습니다.
~~JpaRepository, ~~MongoRepository와 같은 인터페이스를 한번 더 추상화하고, 이를 Service가 알고 있는 구조네요!
확실히 관심사 분리도 되고, Spring Data의 구현체를 활용할 수 있으니 좋은 것 같아요.
해당 방식으로 전부 구조 변경했습니다.

그럼 이번 리뷰도 잘부탁드립니다.😃

#### 애플리케이션 실행 관련
애플리케이션 실행 시 사용할 수 있는admin과 user 계정을 등록해두었습니다. 로그인 정보는 다음과 같습니다.
```
admin
id : admin@gmail.com
password : qwer!

user
id : user@gmail.com
password : qwer!
```

또한 token 관련 민감 정보를 git에 올리지 않게 해두었습니다. 혹시 실행해보신다면 다음과 같이 properties를 추가하셔야합니다!
  

**resources/application-auth.properties**
```
security.jwt.token.secret_key=3C33F72F97777A4EE3F218448AF1DABS
security.jwt.token.expiration_term=3600000
```

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kjyyjk committed:
docs: 예약 대기 기능 요구사항 정의

kjyyjk committed:
refactor: JPA 레포지토리 인터페이스와 구현체 통합

- 단순 위임만 하는 구현체를 만들어 줄 필요가 없어 복잡성이 줄어든다.

- Repository와 JpaRepository 사이에 로직이 존재하는 경우 해당 구조는
사용하기 어렵다.

kjyyjk committed:
feat: 예약 대기 기능 및 예외 구현, 테스트

- 과거 시간 혹은 예약이 없는 경우 대기 요청 시 예외를 발생한다.

kjyyjk committed:
refactor: JpaRepository 인터페이스 패키지 분리

kjyyjk committed:
feat: 예약 대기 추가 기능 구현

kjyyjk committed:
feat: 예약 대기 순번 추가 구현

kjyyjk committed:
feat: 중복 예약 검증 로직 추가 및 테스트

- 동일한 날짜, 테마, 시간, 멤버에 대해 중복 대기가 발생할 경우 예외를 던진다.

kjyyjk committed:
refactor: 테스트 클래스명을 IntegrationTest에서 SliceTest로 변경

kjyyjk committed:
feat: html/js 예약대기 기능 구현

kjyyjk committed:
refactor: waiting 패키지 분리

- Waiting과 Reservation은 서로 다르고, 어느 한 쪽이 다른 쪽의 하위 도메인도 아니라고 판단했다.
따라서 Waiting 관련 클래스를 `waiting`으로 분리했다.

kjyyjk committed:
feat: 예약과 대기 조회 API 구현

kjyyjk committed:
refactor: 복합 도메인 로직 패키지 정리

- Reservation과 Waiting은 도메인 참조 관계보다는 서로 다른 기능의 협력이
필요하다고 판단했다. 따라서 둘 이상의 도메인에 대해 협력 로직을 가지는
파사드 서비스를 도입했다.

- 복합 도메인 로직은 application 패키지에 위치했다.

kjyyjk committed:
feat: 예약 대기 취소 기능 구현

- 본인의 것이 아닌 예약 대기를 취소하면 예외가 발생한다.
- 예약 대기 취소 시 뒤 대기들의 순번이 1씩 앞당겨진다.

kjyyjk committed:
docs: 예약 대기 관리 기능 요구사항 정의

kjyyjk committed:
feat: 예약 대기 관리 html/js 추가

kjyyjk committed:
feat: 어드민의 대기 취소 기능 구현 및 테스트

kjyyjk committed:
feat: 어드민 대기 조회 구현

kjyyjk committed:
feat: 예약 취소 시 대기순번 변경 및 자동 예약 승인 기능구현

kjyyjk committed:
feat: 예약으로 변환된 대기 삭제 구현

kjyyjk committed:
refactor: viewcontroller 제거

kjyyjk committed:
refactor: 테스트 환경 재구성(DirtiesContext 제거)

- 현재 구현에서 스프링 빈은 상태를 가지지 않는다.
따라서 DirtiesContext로 컨텍스트를 새로 로드할 필요가 없어 해당 어노테이션을 제거했다.
다만 테스트 간에 컨텍스트를 공유하는 경우가 있어 db @Sql로 초기 삽입해준 데이터가
다른 테스트에 영향을 줄 수 있다.
따라서 각 테스트에서 컨텍스트를 새로 로드하기 보다는 스키마 초기화와 데이터 삽입을 진행해준다.

kjyyjk committed:
test: Waiting 기능 통합 테스트

kjyyjk committed:
test: 예약&대기 관련 통합 테스트

kjyyjk committed:
feat: 로그아웃 기능 구현

kjyyjk committed:
refactor: ReservationRepository 구현체 삭제

kjyyjk committed:
refactor: 코드래빗 리뷰 반영

kjyyjk committed:
test: WaitingJpaRepository 커스텀 쿼리 테스트

kjyyjk commented on src/main/java/roomescape/waiting/repository/jpa/WaitingJpaRepository.java:
```diff
@@ -0,0 +1,46 @@
+
+public interface WaitingJpaRepository extends JpaRepository<Waiting, Long>, WaitingRepository {
+
+    @Override
+    @Modifying(clearAutomatically = true, flushAutomatically = true)
+    @Query("""
+            UPDATE Waiting w 
+            SET w.priority = (w.priority - :amount)
+            WHERE w.priority >= :from AND w.date = :date AND w.time = :time AND w.theme = :theme
+            """
+    )
+    void pullPriority(
```
순번을 당기는 로직에서 변경 감지를 활용할 경우 update 쿼리가 하나씩 나갈 것을 고려해,
이처럼 JPQL로 직접 update를 해주었습니다.

그리고 `@Query`를 활용한 쓰기 작업은 영속성 컨텍스트와 DB 간 데이터 불일치를 유발할 수 있다고 학습했습니다.
따라서 `@Modifying` 어노테이션과 그 속성들을 활용해주었습니다.
clear 옵션을 사용한 이유는 DB와 불일치하는 컨텍스트를 초기화하기 위함입니다.
flush 옵션을 사용한 이유는 기존에 발생했던 변경 사항을 clear 이전에 DB로 반영하기 위함입니다.

불일치는 해결했지만 영속성 컨텍스트를 초기화했기 때문에 이후 작업에서는 전부 DB까지 쿼리가 나가야한다는
단점도 같이 고려해 볼 필요가 있다고 생각했습니다.

verus-j replied:
현재 트랜잭션을 적용하진 않은 것 같네요. 만약에 트랜잭션을 적용해 여러 쿼리들을 같은 생명주기로 관리하게 된다면 clear나 flush와 같은 옵션이 필요할까요?

kjyyjk replied:
트랜잭션을 적용해 여러 쿼리들을 같은 생명주기로 관리한다는 것이 곧 JPA의 변경 감지를 활용한다는 것으로 이해했습니다.
이번 리뷰 반영하며 변경 감지를 활용하도록 리팩터링 했습니다!
db8789ce781aa9365084a6e38067ec2eda116a0b

혹시 제가 잘못 이해한 것이라면, 다시 언급주시면 감사하겠습니다.😊

verus-j replied:
잘 적용하셨습니다. JPA를 더 잘 활용한것 같네요~

kjyyjk commented on src/main/java/roomescape/waiting/domain/Waiting.java:
```diff
@@ -0,0 +1,71 @@
+    private Theme theme;
+    @OneToOne
+    private Member member;
+    private long priority;
```
대기 순번을 DB에 저장할지 말지에 대해 고민/토론 해보았습니다.
그리고 저는 DB에 저장하기를 선택했습니다.

이유는 다음과 같습니다.
- 방탈출은 도메인 특성 상 캐치테이블처럼 대기가 많이 몰리지도 않으며 취소가 잦지도 않을 것이다. 오히려 순번 조회가 더 잦을 것이다. DB에 순번을 저장하면 조회마다 순번을 연산할 필요가 없다.
- 여러 대기의 순번이 앞당겨지더라도 하나의 update 쿼리로 해결할 수 있다.
- 이후 지난 대기 내역과 같은 기능이 추가됐을 때 순번 정보도 같이 저장할 수 있다.

또한 현재 방식에서 고려해야 할 점은 다음과 같습니다.
- 새로운 예약 대기를 생성할 때 (대기 개수 + 1)로 순번을 계산하고 DB에 삽입을 하는데, 동시성 문제가 발생할 수 있다.
- 동시성 문제를 고려해 락을 걸었을 때 요청이 몰리는 경우 성능 문제가 발생할 수 있다.

현재 단계에서는 고려할 점에 대해 생각만 해봤습니다!

verus-j replied:
대기 순번을 조정하는 로직이 repository에 pullPriority에 있군요. 도메인인 Waiting에서 상태를 변경하여 JPA를 통해 감지된 변경을 DB에 적용하면 안되나요?

kjyyjk replied:
맞습니다!
말씀주신 방법도 고려해봤었지만, 변경 감지를 활용하면 n개의 update 쿼리가 발생한다는 점을 고려해
JPQL과 `@Query`로 bulk update 하기를 선택했습니다.

하지만 이번 리뷰를 받고,
과연 update 쿼리 n번 발생하는 것이 성능에 큰 영향을 미칠까?하고 생각해봤습니다.

JPA는 쓰기 지연 기능으로 여러 건의 쿼리를 처리할 것이기 때문에, 현재 단계에서 그러한 성능을 고려할 필요는 없다고
판단했습니다.
(무엇보다 성능이 좋고 안좋고에 대한 기준이 없기도 합니다ㅎ
bulk update 방식은 보통 1000건 이상의 쿼리를 처리할 때 사용해볼 수 있는 선택지라고 하네요..!)


따라서 변경 감지를 활용하도록 리팩터링했습니다. 
관련된 개념으로 batch size에 대해서도 추후 학습해보겠습니다.
db8789ce781aa9365084a6e38067ec2eda116a0b

kjyyjk commented on src/main/java/roomescape/application/reservationwaiting/service/ReservationWaitingService.java:
```diff
@@ -0,0 +1,58 @@
+import roomescape.waiting.service.WaitingService;
+import roomescape.waiting.service.dto.WaitingInfo;
+
+@Transactional
+@Service
+public class ReservationWaitingService {
+
+    private final ReservationService reservationService;
+    private final WaitingService waitingService;
```
`Reservation`과 `Waiting`을 서로 다른 독립적인 도메인으로 바라보고 있기 때문에,
두 도메인의 복합 비즈니스 로직은 `ReservationWaitingService`라는 별도의 Service를 만들어 처리하도록 했습니다.

verus-j replied:
복합 비즈니스 로직을 별도로 관리한다면 앞으로 여러 도메인 클래스가 복합적으로 처리해야하는 로직이 생기면 별도로 계속해서 Service를 추출할 계획일까요?

kjyyjk replied:
현재 방식으로 구현하며 걱정했던 부분이기도 합니다.
n개의 도메인이 각각 독립적인 비즈니스 로직을 가지고 있다면 2^n개의 서비스를 만들어야 할 수도 있을 것 같아요.

이러한 구조를 가져간 이유 중 한 가지는 의존성 관리였습니다.
예약 대기라는 기능은 충분히 사라질 수 있는 것이기에, 예약 기능은 예약 대기 기능으로부터 온전해야 한다고 생각했습니다.

그렇기 때문에 ReservationService가 WaitingService에 의존하게 된다면 제 의도와 달리 앞뒤가 안맞는다고 생각했습니다.
그렇다고 ReservationService가 WaitingRepository를 사용해 직접 조작하는 것은 과한 책임을 갖는다고 생각했습니다.
이미 WaitingService에 있는 로직이 ReservationService에 중복으로 생길 수도 있을 것 같고요.

따라서 서비스 수가 많아질 수도 있지만 의존성 문제를 풀어낼 수 있을 것 같아 이 같은 파사드 방식을 적용했습니다.

베루스는 이에 대해 어떻게 생각하시는지 궁금합니다!
아직 의존성의 흐름이 얼마나 중요하고 어떻게 잘 제어할 수 있는지 잘 모르겠기도 합니다..!

외에 이벤트 발행 방식을 생각해보기도 했지만,
직관적이지 않다는 단점이 현재 단계에서의 장점보다 크다고 판단해 선택하지 않았습니다.

verus-j replied:
위와 같이 퍼사드를 만드는 것이 ReservationService가 WaitingRepository를 직접 조작하는 것보다 심플한지 잘 모르겠습니다. 퍼사드를 너무 이르게 적용했다는 생각이 들고, 그로인해 너무 세밀하게 코드를 분리하는 느낌이 드는 것 같습니다. 오히려 지금 구조에서는 Repository를 여러 서비스에서 재사용하는게 심플한 방법으로 보입니다. 아직 개인적으로는 과한 책임이라고 생각이 들지 않네요 ㅠㅠ

그래도 레오가 미션을 진행하면서 과한 책임을 갖는다고 판단하셨으니 제 개인적인 의견만 남기고 여기서 마무리하겠습니다! 아직 미션이 하나 더 남아 있으니 진행하면서 퍼사드 방식에 대해 계속 고민해보면 좋을 것 같아요~

kjyyjk replied:
미션을 거듭하면서 책임 분리에 과몰입한 나머지, 트레이드 오프를 고려하지 못한 것 같아요!
말씀주신 것처럼 과연 현재 필요한 방식인지 장단점을 고려해 도입해봐야겠어요.

피드백 감사합니다~!

kjyyjk reviewed (commented):
리뷰 감사합니다!

verus-j commented on src/main/java/roomescape/application/reservationwaiting/controller/ReservationWaitingController.java:
```diff
@@ -0,0 +1,49 @@
+                myReservationAndWaiting.waitingInfos()
+                        .stream()
+                        .map(MyReservationAndWaitingResponse::new)
+        ).toList();
```
concat 내부의 로직을 변수로 추출해 가독성을 챙길 수 있을 것 같습니다.

kjyyjk replied:
수정했습니다.
2bd94db3dc52c2e72d7a57183e09874b25624caf

verus-j commented on src/test/java/roomescape/integration/AdminRoleInterceptorMockMvcTest.java:
```diff
@@ -21,16 +19,15 @@
 @Transactional
-@Sql(scripts = "/member-data.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_CLASS)
+@Sql(scripts = {"/schema.sql", "/member-data.sql"}, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_CLASS)
 public class AdminRoleInterceptorMockMvcTest {
```
테스트가 실패하네요... 혹시 따로 제가 설정해야하는 부분이 있는걸까요?
아래와 같이 MemberRepository 빈을 찾을 수 없어 에러가 발생하고 있습니다.

![스크린샷 2025-05-27 오전 1 07 09](https://github.com/user-attachments/assets/0f81c0c4-c6ab-4a37-b6d1-9bccf1e7a52e)

kjyyjk replied:
`application-auth.properties` 외에는 전부 git으로 관리하고 있고,
제 환경에서는 문제 없이 전체 테스트를 통과하고 있어서 왜 이런지 잘 모르겠습니다...ㅠ

혹시 특정 테스트가 아닌 전체 테스트에서 실패하는 것일까요?

verus-j replied:
제 로컬 세팅이 문제가 있나봅니다 ㅠㅠ 레오 로컬에서는 모두 통과하고 있으니 넘어가시죠!

verus-j commented on src/test/java/roomescape/integration/ReservationWaitingMockMvcSpringBootTest.java:
```diff
@@ -0,0 +1,143 @@
+        // then
+        mockMvc.perform(get("/reservations-mine").cookie(cookie))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.length()").value(3));
```
사이즈 비교만으로 충분할까요?

kjyyjk replied:
테스트 목적에 따라 사이즈 비교도 충분하겠지만 제가 작성한 대부분의 테스트는 그렇지 않은 것 같습니다.
따라서 해당하는 테스트들에 응답 json의 상태 검증문도 추가했습니다!
bb3f960fa5b4f9a92e94823bebdbc482bacf837b

verus-j reviewed (changes requested):
안녕하세요 레오
미션 진행하느라 고생하셨습니다.
피드백 남겨뒀으니 확인부탁드려요~

kjyyjk committed:
refactor: 대기 순번 당길 때 변경감지를 활용하도록 리팩터링

- 기존에는 JPQL을 활용한 bulk update로 순번들을 당겨주고 있었다.
그리고 `@Modifying` 어노테이션과 clear, flush 옵션을 사용해
데이터베이스와 영속성 컨텍스트 간의 불일치 문제를 해결하고 있었다.
다만 순번을 조정하는 것 자체가 비즈니스 로직인데 쿼리에 담겨있다는 점,
영속성 컨텍스트를 초기화하면 캐싱을 활용하지 못하고 전부 db까지 찔러야한다는 문제가 생긴다.
비록 변경 감지를 활용하면 n번의 update 쿼리가 발생하지만, jpa의 지연 쓰기 기능으로 인해 큰 문제가 없을 것이라 판단했다.

- 변경 감지를 활용하도록 리팩터링한 뒤, bulk update는 추후 성능에 이슈가 생기면 최적화를 목적으로 도입할 수 있다.

kjyyjk committed:
refactor: 내 예약 정보 조회 로직 리팩터링

kjyyjk committed:
refactor: 테스트 시 사이즈 외 상태 검증 추가

verus-j reviewed (approved):
안녕하세요 레오
피드백 반영해주신것 확인했습니다!
미션 진행하느라 고생 많으셨고, 여기서 이만 머지하겠습니다~
